scaler/io/ymq/message_connection_tcp.cpp

```diff
@@ -0,0 +1,305 @@
+        if (iovecs.size() > IOV_MAX - 2) {
+            break;
+        }
+
+        iovec iovHeader {};
+        iovec iovPayload {};
+        if (it == _writeOperations.begin()) {
+            if (_sendCursor < HEADER_SIZE) {
+                iovHeader.iov_base  = (char*)(&it->_header) + _sendCursor;
+                iovHeader.iov_len   = HEADER_SIZE - _sendCursor;
+                iovPayload.iov_base = (void*)(it->_payload.data());
+                iovPayload.iov_len  = it->_payload.len();
+            } else {
+                iovHeader.iov_base  = nullptr;
+                iovHeader.iov_len   = 0;
+                iovPayload.iov_base = (char*)(it->_payload.data()) + (_sendCursor - HEADER_SIZE);
+                iovPayload.iov_len  = it->_payload.len() + (_sendCursor - HEADER_SIZE);
+            }
+        } else {
+            iovHeader.iov_base  = (void*)(&it->_header);
+            iovHeader.iov_len   = HEADER_SIZE;
+            iovPayload.iov_base = (void*)(it->_payload.data());
+            iovPayload.iov_len  = it->_payload.len();
+        }
+
+        iovecs.push_back(iovHeader);
+        iovecs.push_back(iovPayload);
+    }
+
+    if (iovecs.empty()) {
+        return 0;
+    }
+
+    struct msghdr msg {};
+    msg.msg_iov    = iovecs.data();
+    msg.msg_iovlen = iovecs.size();
+
+    ssize_t bytesSent = ::sendmsg(_connFd, &msg, MSG_NOSIGNAL);
+    if (bytesSent == -1) {
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+            return 0;
+        } else
+            return std::unexpected {errno};
+    }
+
+    return bytesSent;
+}
+
+// TODO: There is a classic optimization that can (and should) be done. That is, we store
+// prefix sum in each write operation, and perform binary search instead of linear search
+// to find the first write operation we haven't complete. - gxu
+void MessageConnectionTCP::updateWriteOperations(size_t n) {
+    auto target = _writeOperations.begin();
+    _sendCursor += n;
+    // Post condition of the loop: target contains the first write op we haven't complete.
+    for (auto it = _writeOperations.begin(); it != _writeOperations.end(); ++it) {
+        size_t msgSize = it->_payload.len() + HEADER_SIZE;
+        if (_sendCursor < msgSize) {
+            target = it;
+            break;
+        }
+
+        if (_sendCursor == msgSize) {
+            target      = it + 1;
+            _sendCursor = 0;
+            break;
+        }
+
+        _sendCursor -= msgSize;
+    }
+
+    // NOTE: [complete, _nwriteOperations.end()) is a range of completed write operations.
+    // After rotate, the _writeOperations contains something like this:
+    //  [Incomplete0, Incomplete1, ..., complete, complete1, ..., completeN]
+    auto complete = std::rotate(_writeOperations.begin(), target, _writeOperations.end());
+
+    for (auto tmp = complete; tmp != _writeOperations.end(); ++tmp) {
+        tmp->_callbackAfterCompleteWrite(0);
+    }
+
+    _writeOperations.erase(complete, _writeOperations.end());
```

Comment:
I think you might be able to simplify this logic using a `std::queue` instead of `std::vector`:

```suggestion
    // typeof(_writeOperations) == std::vector<WriteOperation>
    
    // Cleanup completed operations
    while (writeOperation.front() != target) {
         _writeOperations.front()->_callbackAfterCompleteWrite(0);
         _writeOperations.pop();
    }
```